pkg/3scale/amp/operator/upgrade.go

```diff
@@ -727,6 +736,72 @@ func (u *UpgradeApiManager) ensurePodTemplateLabels(desired *appsv1.DeploymentCo
 	return updated, nil
 }
 
+func (u *UpgradeApiManager) upgradePVCSizes() (reconcile.Result, error) {
+	if !u.apiManager.IsExternalDatabaseEnabled() {
+		if u.apiManager.IsSystemPostgreSQLEnabled() {
+			err := u.updatePVCSize(component.SystemPostgreSQLPVCName, resource.MustParse("10Gi"))
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
+		if u.apiManager.IsSystemMysqlEnabled() {
+			err := u.updatePVCSize(component.SystemMysqlPVCName, resource.MustParse("10Gi"))
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
+		// Backend redis PVC
+		err := u.updatePVCSize(component.BackendRedisStorageVolumeName, resource.MustParse("10Gi"))
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// System redis PVC
+		err = u.updatePVCSize(component.SystemRedisStorageVolumeName, resource.MustParse("10Gi"))
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	if !u.apiManager.IsSystemS3Enabled() {
+		err := u.updatePVCSize(component.SystemFileStoragePVCName, resource.MustParse("10Gi"))
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (u *UpgradeApiManager) updatePVCSize(pvcName string, desiredSize resource.Quantity) error {
+	existing := &v1.PersistentVolumeClaim{}
+	namespacedName := types.NamespacedName{
+		Name:      pvcName,
+		Namespace: u.apiManager.Namespace,
+	}
+	err := u.Client().Get(context.TODO(), namespacedName, existing)
+	// NotFound also regarded as error, as pvc is expected to exist
+	if err != nil {
+		return err
+	}
+
+	// v1.ResourceStorage must exist, otherwise PVC is regarded as invalid
+	existingSize := existing.Spec.Resources.Requests[v1.ResourceStorage]
+
+	if existingSize.Cmp(desiredSize) != 0 {
```

Comment:
I seem to remember there were some comparison methods that checked unexported fields, and that made detecting differences due to different internal representations of the same value depending on the serialized value. Have we checked this is working as expected in all cases?

That's why helper.CmpResources was created in its day for other kind of resources. I don't know whether CmpResources takes into account ResourceStorage though

Reply:
CmpResources does not take into account ResourceStorage. Already checked it. 

The upgrade logic uses `Quantity` type comparator, which I expect to be working

```
func main() {
	desired := resource.MustParse("10Gi")
	existing := resource.MustParse("1Gi")
	fmt.Println("desired == existing", existing.Cmp(desired) == 0)
	existing = resource.MustParse("10Gi")
	fmt.Println("desired == existing", existing.Cmp(desired) == 0)
}
```

```
desired == existing false
desired == existing true
```